engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/ThrowPanicNode.java

```diff
@@ -1,16 +1,88 @@
 package org.enso.interpreter.node.expression.builtin.error;
 
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.BranchProfile;
 import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.callable.atom.Atom;
 import org.enso.interpreter.runtime.error.PanicException;
 import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(
     type = "Panic",
     name = "throw",
     description = "Throws a new Panic with given payload.")
-public class ThrowPanicNode extends Node {
-  Stateful execute(Object _this, Object payload) {
+public abstract class ThrowPanicNode extends Node {
+  static ThrowPanicNode build() {
+    return ThrowPanicNodeGen.create();
+  }
+
+  abstract Stateful execute(Object _this, Object payload);
+
+  Context getContext() {
+    return Context.get(this);
+  }
+
+  @Specialization(
+      guards = {"payload.getConstructor() == getContext().getBuiltins().caughtPanic()"})
+  Stateful doCaughtPanic(
+      Object _this,
+      Atom payload,
+      @CachedLibrary(limit = "5") InteropLibrary interopLibrary,
+      @Cached BranchProfile typeErrorProfile) {
+    // Note [Original Exception Type]
+    Object originalException = payload.getFields()[1];
+    if (interopLibrary.isException(originalException)) {
+      try {
+        throw interopLibrary.throwException(originalException);
+      } catch (UnsupportedMessageException e) {
```

Comment:
please blow up with `IllegalStateException` – we shouldn't want something that deeply broken to "kinda work" – I'd rather we inform people that this is a bug.